File: macclean/cleaners/spotlight.py
from pathlib import Path
import click
from rich.console import Console
from rich.panel import Panel
from macclean.core.utils import AnalysisResult, confirm, run_cmd, require_sudo

console = Console()


def analyze(home: Path | None = None) -> AnalysisResult:
    return AnalysisResult()


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    console.print(Panel("[bold cyan]Spotlight Reindex[/]"))
    console.print("  This will reindex the entire drive (Spotlight unavailable for ~10 min).")

    if dry_run:
        return
    if not yes and not confirm("Reindex Spotlight?"):
        return

    require_sudo()
    out, code = run_cmd(["mdutil", "-E", "/"], timeout=30)
    if code == 0:
        console.print("  [green]+[/] Spotlight reindex started")
    else:
        console.print(f"  [yellow]![/] mdutil error: {out}")


@click.command()
@click.pass_context
def cmd(ctx):
    """Reindex Spotlight (sudo mdutil -E /)."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/spotlight.rs
use anyhow::Result;
use crate::core::{AnalysisResult, Cleaner};
use crate::core::cmd::{run_cmd, require_sudo};
use crate::ui;

pub struct SpotlightCleaner;

impl Cleaner for SpotlightCleaner {
    fn name(&self) -> &str { "spotlight" }
    fn display_name(&self) -> &str { "Spotlight" }

    fn analyze(&self) -> Result<AnalysisResult> {
        Ok(AnalysisResult::default())
    }

    fn clean(&self, _result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        println!("  This will reindex the entire drive (Spotlight unavailable for ~10 min).");
        if dry_run { return Ok(()); }
        if !yes && !ui::confirm("Reindex Spotlight?", false)? { return Ok(()); }

        require_sudo();
        let r = run_cmd(&["mdutil", "-E", "/"]);
        if r.success() {
            ui::print_ok("Spotlight reindex started.");
        } else {
            ui::print_warn(&format!("mdutil failed: {}", &r.output[..r.output.len().min(200)]));
        }
        Ok(())
    }
}
